src/ray/core_worker/gcs_server_address_updater.cc

```diff
@@ -0,0 +1,64 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/gcs_server_address_updater.h"
+
+namespace ray {
+
+GcsServerAddressUpdater::GcsServerAddressUpdater(
```

Comment:
Why don't we just use the core worker io service instead of creating one for this?

Reply:
+1, I have the same question. You could then use the existing periodical runner and raylet client on the core worker object, running `UpdateGcsServerAddress` + the `update_func` callback as the periodic function.

Reply:
I'd like to update GCS server address with separate thread as it might be impacted(even blocked) by main thread.
The reason I don't reuse the raylet client in the core worker object is that we can create this gcs client in core worker process and reuse it in core workers. We've already done like this internally and will create a PR to synchronize it to community in next PR.

Reply:
> this gcs client in core worker process and reuse it in core workers.

Is it to reduce the number of connections by having 1 connection per process instead of core worker threads for Java? What's the purpose?

Reply:
> I'd like to update GCS server address with separate thread as it might be impacted(even blocked) by main thread.

Also, I don't think this is a big deal. Did you guys discover issues from this scenario?

Reply:
> > this gcs client in core worker process and reuse it in core workers.
> 
> Is it to reduce the number of connections by having 1 connection per process instead of core worker threads for Java? What's the purpose?

Yes if we reuse gcs client in core worker threads, the number of connections to gcs will be reduced.

Reply:
> > I'd like to update GCS server address with separate thread as it might be impacted(even blocked) by main thread.
> 
> Also, I don't think this is a big deal. Did you guys discover issues from this scenario?

Actually in the first version I implemented the gcs address udater uses main thread, but we found it will be blocked by retrying connect logic in gcs client. The updater would wait for main thread while the reconnect logic wait for valid gcs address, which causes dead lock.

Before @kfstorm and I also have disagreement about using separate thread or the main on, just as you and me now. :-p But the truth shows separate thread is a better choice.

Reply:
Got it, that makes sense. Thanks @WangTaoTheTonic!